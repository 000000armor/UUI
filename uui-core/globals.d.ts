/// <reference path="../uui_globals.d.ts" />

declare let __PACKAGE_VERSION__: string;

declare module 'jsurl';
declare module 'query-string';
