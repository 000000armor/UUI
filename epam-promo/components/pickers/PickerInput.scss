@use  '../../assets/styles/index' as *;

.footer-wrapper {
    border-top: 1px solid $gray40;
    background: white;
    padding: 0 12px;
    display: flex;
}

.footer-size-24 {
    min-height: 30px;
    padding: 0 12px;
}

.footer-size-30 {
    min-height: 30px;
}

.footer-size-36 {
    min-height: 36px;

    :global(.uui-caption) {
        font-size: 14px;
    }
}

.footer-size-42 {
    min-height: 42px;

    :global(.uui-caption) {
        font-size: 16px;
    }
}

.footer-size-48 {
    min-height: 48px;

    :global(.uui-caption) {
        font-size: 16px;
    }
}
