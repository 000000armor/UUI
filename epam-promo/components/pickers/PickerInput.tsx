import React from 'react';
import { DataRowProps, IDropdownToggler, IEditableDebouncer, isMobile, uuiMarkers } from '@epam/uui';
import { DropdownBodyProps, PickerInputBase, PickerTogglerProps } from '@epam/uui-components';
import { PickerModal } from './PickerModal';
import { Panel } from '../layout';
import { PickerToggler, PickerTogglerMods } from './PickerToggler';
import { DataPickerRow } from './DataPickerRow';
import { PickerItem } from './PickerItem';
import { DataPickerBody } from './DataPickerBody';
import { DataPickerFooter } from './DataPickerFooter';
import { MobileDropdownWrapper } from './MobileDropdownWrapper';
import { EditMode, IHasEditMode, SizeMod } from '../types';
import css from './PickerInput.scss';

export type PickerInputProps = SizeMod & IHasEditMode & {};

const pickerHeight = 300;
const pickerWidth = 360;

export class PickerInput<TItem, TId> extends PickerInputBase<TItem, TId, PickerInputProps> {
    toggleModalOpening(opened: boolean) {
        this.context.uuiModals.show(props => <PickerModal<TItem, TId>
            { ...this.props }
            { ...props }
            caption={ this.getPlaceholder() }
            initialValue={ this.props.value as any }
            renderRow={ this.renderRow }
            selectionMode={ this.props.selectionMode }
            valueType={ this.props.valueType }
        />)
            .then(newSelection => this.handleSelectionValueChange(newSelection))
            .catch(() => null);
    }

    getRowSize() {
        return isMobile()
            ? '48'
            : this.props.editMode === 'modal'
                ? '36'
                : this.props.size;
    }

    renderItem = (item: TItem, rowProps: DataRowProps<TItem, TId>) => {
        return <PickerItem title={ this.getName(item) } size={ this.getRowSize() } { ...rowProps } />;
    }

    renderRow = (rowProps: DataRowProps<TItem, TId>) => {
        if (rowProps.isSelectable && this.isSingleSelect() && this.props.editMode !== 'modal') {
            rowProps.onSelect = this.onSelect;
        }

        return this.props.renderRow ? this.props.renderRow(rowProps) : (
            <DataPickerRow
                { ...rowProps }
                key={ rowProps.rowKey }
                borderBottom='none'
                size={ this.getRowSize() }
                rawProps={ { 'aria-selected': rowProps.isSelectable && rowProps.isSelected, role: 'option' } }
                padding={ this.props.editMode === 'modal' ? '24' : '12' }
                renderItem={ this.renderItem }
            />
        );
    }

    getTogglerProps(rows: DataRowProps<TItem, TId>[]): PickerTogglerProps<TItem, TId> & PickerTogglerMods {
        return {
            ...super.getTogglerProps(rows),
            size: this.props.size as PickerTogglerMods['size'],
            mode: this.props.mode ? this.props.mode : EditMode.FORM,
        };
    }

    renderFooter() {
        const view = this.getView();

        return this.props.renderFooter
            ? this.props.renderFooter({
                ...this.props as any,
                view: view,
                showSelected: {
                    value: this.state.showSelected,
                    onValueChange: (nV) => this.setState({ showSelected: nV, dataSourceState: { ...this.state.dataSourceState } }),
                },
            })
            : (
                <DataPickerFooter
                    isSingleSelect={ this.isSingleSelect() }
                    size={ this.props.size }
                    hasSelection={ view.getSelectedRows().length > 0 }
                    clearSelection={ this.clearSelection }
                    switchValue={ this.state.showSelected }
                    onSwitchValueChange={ (nV) => this.setState({ showSelected: nV }) }
                    selectAll={ view.selectAll }
                />
            );
    }

    renderTarget(dropdownProps: IDropdownToggler): React.ReactNode {
        const rows = this.getRows();
        const renderTarget = this.props.renderToggler || (props => <PickerToggler { ...props } />);

        return (
            <IEditableDebouncer
                value={ this.state.dataSourceState.search }
                onValueChange={ this.handleTogglerSearchChange }
                render={ editableProps => renderTarget({ ...this.getTogglerProps(rows), ...dropdownProps, ...editableProps }) }
            />
        );
    }

    renderBody(props: DropdownBodyProps) {
        const rows = this.getRows();
        const renderedDataRows = rows.map((props: DataRowProps<TItem, TId>) => this.renderRow(props));

        const maxHeight = isMobile()
            ? document.documentElement.clientHeight
            : (this.props.dropdownHeight || pickerHeight);
        const minBodyWidth = isMobile()
            ? document.documentElement.clientWidth 
            : (this.props.minBodyWidth || pickerWidth);
        
        return (
            <Panel
                shadow
                style={ { width: props.togglerWidth > minBodyWidth ? props.togglerWidth : minBodyWidth } }
                rawProps={ { tabIndex: -1 } }
                cx={ [css.panel, uuiMarkers.lockFocus] }
            >
                <MobileDropdownWrapper
                    title={ this.props.entityName }
                    close={ () => this.toggleBodyOpening(false) }
                >
                    <DataPickerBody
                        { ...this.getListProps() }
                        { ...this.getPickerProps() }
                        rows={ renderedDataRows }
                        maxHeight={ maxHeight }
                        onKeyDown={ (e: React.KeyboardEvent<HTMLElement>) => this.handlePickerInputKeyboard(rows, e) }
                        scheduleUpdate={ props.scheduleUpdate }
                        searchSize={ this.props.size }
                        editMode='dropdown'
                    />
                    { this.renderFooter() }
                </MobileDropdownWrapper>
            </Panel>
        );
    }
}
