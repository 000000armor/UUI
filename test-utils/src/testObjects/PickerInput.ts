import { fireEvent, within, screen } from '../extensions/testingLibraryReactExt';

export class PickerInputTestObject {
    static getOptions(props: { busy?: boolean, editMode?: string } = {}) {
        const dialog = within(this.getDialog(props.editMode));
        const params: any = {};
        if (typeof props.busy !== 'undefined') {
            params.busy = props.busy;
        }
        return dialog.getAllByRole('option', params);
    }

    static async findOptions(props: { busy?: boolean, editMode?: string } = {}) {
        const dialog = within(await this.findDialog(props.editMode));
        const params: any = {};
        if (typeof props.busy !== 'undefined') {
            params.busy = props.busy;
        }
        return await dialog.findAllByRole('option', params);
    }

    static async findOptionsText(props: { busy?: boolean, editMode?: string } = {}) {
        const opts = await this.findOptions(props);
        return opts.map((o) => o.textContent?.trim());
    }

    static async findCheckedOptions() {
        const dialog = within(await this.findDialog());
        return (await dialog.findAllByRole('option')).filter((opt) => {
            return (within(opt).getByRole('checkbox') as HTMLInputElement).checked;
        }).map((e) => e.textContent?.trim());
    }

    static async findUncheckedOptions() {
        const dialog = within(await this.findDialog());
        return (await dialog.findAllByRole('option')).filter((opt) => {
            return !(within(opt).getByRole('checkbox') as HTMLInputElement).checked;
        }).map((e) => e.textContent?.trim());
    }

    static getPlaceholderText(input: HTMLElement) {
        return input.getAttribute('placeholder')?.trim();
    }

    static clearInput(container: HTMLElement) {
        const clearButton = within(container).getByRole('button', { name: 'Clear' });
        fireEvent.click(clearButton);
    }

    static hasClearInputButton(container: HTMLElement) {
        return !!within(container).queryByRole('button', { name: 'Clear' });
    }

    static async clickSelectAllOptions() {
        const dialog = within(await this.findDialog());
        const selectAllButton = await dialog.findByRole('button', { name: 'SELECT ALL' });
        fireEvent.click(selectAllButton);
    }

    static async clickShowOnlySelected() {
        const dialog = within(await this.findDialog());
        const showOnlySelected = await dialog.findByRole('switch', { name: 'Show only selected' });
        fireEvent.click(showOnlySelected);
    }

    static async clickClearAllOptions() {
        const dialog = within(await this.findDialog());
        const selectAllButton = await dialog.findByRole('button', { name: 'CLEAR ALL' });
        fireEvent.click(selectAllButton);
    }

    static getSelectedTagsText(input: HTMLElement) {
        return this.getSelectedTags(input).map((b) => b.textContent?.trim());
    }

    static removeSelectedTagByText(input: HTMLElement, text: string) {
        const tag = this.getSelectedTags(input).find((b) => b.textContent?.trim() === text);
        const removeTagIcon = tag?.lastElementChild;
        fireEvent.click(removeTagIcon as Element);
    }

    static async clickOptionByText(optionText: string) {
        const opt = await this.findOption(optionText);
        fireEvent.click(opt);
    }

    static async clickOptionCheckbox(optionText: string) {
        const opt = await this.findOption(optionText);
        fireEvent.click(within(opt).getByRole('checkbox'));
    }

    static async clickOptionUnfold(optionText: string) {
        const opt = await this.findOption(optionText);
        fireEvent.click(within(opt).getByRole('button', { name: 'Unfold' }));
    }

    static async hasOptions(props?: { busy?: boolean }) {
        return (await this.findOptions(props)).length > 0;
    }

    private static async findDialog(editMode: string = 'dialog') {
        return await screen.findByRole(editMode === 'modal' ? 'modal' : 'dialog');
    }

    private static getDialog(editMode: string = 'dialog') {
        return screen.getByRole(editMode === 'modal' ? 'modal' : 'dialog');
    }

    private static async findOption(optionText: string) {
        const dialog = within(await this.findDialog());
        return await dialog.findByRoleAndText({ role: 'option', text: optionText });
    }

    private static getSelectedTags(input: HTMLElement) {
        const tags: HTMLElement[] = [];
        let s = input;
        while ((s = s.previousElementSibling as HTMLElement)) {
            if (s.tagName.toLowerCase() === 'button') {
                tags.push(s);
            }
        }
        return tags.reverse();
    }
}
