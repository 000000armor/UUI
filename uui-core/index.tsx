export * from './constants';
export * from './helpers';
export * from './services';
export * from './types';
export * from './data';
export * from './hooks';
export * from './i18n';
