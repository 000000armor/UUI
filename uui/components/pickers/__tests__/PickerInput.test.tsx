import React, { ReactNode } from 'react';
import { ArrayDataSource, AsyncDataSource, CascadeSelection } from '@epam/uui-core';
import {
    renderSnapshotWithContextAsync, setupComponentForTest, screen, within, fireEvent, delay, waitFor,
    queryHelpers,
} from '@epam/uui-test-utils';
import { PickerInput, PickerInputProps } from '../PickerInput';
import { PickerInputBaseProps } from '@epam/uui-components';

jest.mock('react-popper', () => ({
    ...jest.requireActual('react-popper'),
    Popper: function PopperMock({ children }: any) {
        return children({
            ref: jest.fn,
            update: jest.fn(),
            style: {},
            arrowProps: { ref: jest.fn },
            placement: 'bottom-start',
            isReferenceHidden: false,
        });
    },
}));

type TestItemType = {
    id: number;
    level: string;
    name: string;
};

const languageLevels: TestItemType[] = [
    { id: 2, level: 'A1', name: 'Elementary' },
    { id: 3, level: 'A1+', name: 'Elementary+' },
    { id: 4, level: 'A2', name: 'Pre-Intermediate' },
    { id: 5, level: 'A2+', name: 'Pre-Intermediate+' },
    { id: 6, level: 'B1', name: 'Intermediate' },
    { id: 7, level: 'B1+', name: 'Intermediate+' },
    { id: 8, level: 'B2', name: 'Upper-Intermediate' },
    { id: 9, level: 'B2+', name: 'Upper-Intermediate+' },
    { id: 10, level: 'C1', name: 'Advanced' },
    { id: 11, level: 'C1+', name: 'Advanced+' },
    { id: 12, level: 'C2', name: 'Proficiency' },
];

type TestTreeItem = {
    id: number;
    name: string;
    parentId?: number;
};

const treeLikeData: TestTreeItem[] = [
    { id: 1, name: 'Parent 1' },
    { id: 1.1, parentId: 1, name: 'Child 1.1' },
    { id: 1.2, parentId: 1, name: 'Child 1.2' },
    { id: 1.3, parentId: 1, name: 'Child 1.3' },
    { id: 2, name: 'Parent 2' },
    { id: 2.1, parentId: 2, name: 'Child 2.1' },
    { id: 2.2, parentId: 2, name: 'Child 2.2' },
    { id: 2.3, parentId: 2, name: 'Child 2.3' },
    { id: 3, name: 'Parent 3' },
    { id: 3.1, parentId: 3, name: 'Child 3.1' },
    { id: 3.2, parentId: 3, name: 'Child 3.2' },
    { id: 3.3, parentId: 3, name: 'Child 3.3' },
];

const mockDataSource = new ArrayDataSource({
    items: languageLevels,
});

const mockDataSourceAsync = new AsyncDataSource({
    api: async () => {
        await delay(100);
        return languageLevels;
    },
});

const mockTreeLikeDataSourceAsync = new AsyncDataSource<TestTreeItem, number, any>({
    api: async () => {
        await delay(100);
        return treeLikeData;
    },
    getParentId: ({ parentId }) => parentId,
});

type PickerInputComponentProps<TItem, TId> = PickerInputBaseProps<TItem, TId> & PickerInputProps;

async function setupPickerInputForTest<TItem = TestItemType, TId = number>(params: Partial<PickerInputComponentProps<TItem, TId>>) {
    const { result, mocks } = await setupComponentForTest<PickerInputComponentProps<TItem, TId>>(
        (context): PickerInputComponentProps<TItem, TId> => {
            if (params.selectionMode === 'single') {
                return Object.assign({
                    onValueChange: jest.fn().mockImplementation((newValue) => context.current?.setProperty('value', newValue)),
                    dataSource: mockDataSourceAsync,
                    disableClear: false,
                    searchPosition: 'input',
                    getName: (item) => item.level,
                    value: params.value as TId,
                    selectionMode: 'single',
                }, params) as PickerInputComponentProps<TItem, TId>;
            }

            return Object.assign({
                onValueChange: jest.fn().mockImplementation((newValue) => context.current?.setProperty('value', newValue)),
                dataSource: mockDataSourceAsync,
                disableClear: false,
                searchPosition: 'input',
                getName: (item) => item.level,
                value: params.value as number[],
                selectionMode: 'multi',
            }, params) as PickerInputComponentProps<TItem, TId>;
        },
        (props) => <PickerInput { ...props } />,
    );
    const input = screen.queryByRole('textbox');

    return {
        result,
        mocks,
        dom: { input },
    };
}

describe('PickerInput', () => {
    it('should render with minimum props', async () => {
        const tree = await renderSnapshotWithContextAsync(
            <PickerInput value={ null } onValueChange={ jest.fn } selectionMode="single" dataSource={ mockDataSource } disableClear searchPosition="input" />,
        );
        expect(tree).toMatchSnapshot();
    });

    it('should render with maximum props', async () => {
        const tree = await renderSnapshotWithContextAsync(
            <PickerInput
                value={ [2, 3] }
                onValueChange={ jest.fn }
                selectionMode="multi"
                dataSource={ mockDataSource }
                size="48"
                maxItems={ 20 }
                editMode="modal"
                valueType="id"
                getName={ (item) => item.level }
                autoFocus
                placeholder="Test placeholder"
                filter={ (item: any) => item.level === 'A1' }
                sorting={ { direction: 'desc', field: 'level' } }
                searchPosition="body"
                minBodyWidth={ 900 }
                renderNotFound={ () => null }
                renderFooter={ (props) => <div>{props as unknown as ReactNode}</div> }
                cascadeSelection
                dropdownHeight={ 48 }
                minCharsToSearch={ 4 }
            />,
        );
        expect(tree).toMatchSnapshot();
    });

    it('[selectionMode multi] should select & clear several options', async () => {
        const { dom, mocks } = await setupPickerInputForTest({
            value: undefined,
            selectionMode: 'multi',
        });
        expect(dom.input.getAttribute('placeholder').trim()).toEqual('Please select');
        fireEvent.click(dom.input);
        expect(screen.getByRole('dialog')).toBeInTheDocument();

        const optionC2 = await screen.findByText('C2');
        expect(optionC2).toBeInTheDocument();

        const [cb1, cb2] = await within(screen.getByRole('dialog')).findAllByRole('checkbox');
    
        fireEvent.click(cb1);
        expect(mocks.onValueChange).toHaveBeenLastCalledWith([2]);
        fireEvent.click(cb2);
        expect(mocks.onValueChange).toHaveBeenLastCalledWith([2, 3]);
        expect(cb1).toBeChecked();
        expect(cb2).toBeChecked();
        fireEvent.click(window.document.body);
        expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
        expect(screen.queryAllByRole('button')).toHaveLength(3); // 2 tags and 1 clear button
        const a1 = screen.getByText('A1');
        const a1Clear = a1.nextElementSibling;
        const a1plus = screen.getByText('A1+');
        const a1plusClear = a1plus.nextElementSibling;
        fireEvent.click(a1Clear);
        expect(screen.queryAllByRole('button')).toHaveLength(2); // 1 tag and 1 clear button
        fireEvent.click(a1plusClear);
        expect(screen.queryAllByRole('button')).toHaveLength(0);
    });

    it('[selectionMode single] should select & clear option', async () => {
        const { dom, mocks } = await setupPickerInputForTest({
            value: undefined,
            selectionMode: 'single',
        });
        expect(dom.input.getAttribute('placeholder').trim()).toEqual('Please select');
        fireEvent.click(dom.input);
        expect(screen.getByRole('dialog')).toBeInTheDocument();
        const optionC2 = await screen.findByText('C2');
        fireEvent.click(optionC2);
        expect(mocks.onValueChange).toHaveBeenLastCalledWith(12);
        fireEvent.click(window.document.body);
        expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
        expect(screen.getByPlaceholderText('C2')).toBeInTheDocument();
        const clear = screen.getByRole('button');
        fireEvent.click(clear);
        expect(screen.queryByText('C2')).not.toBeInTheDocument();
    });
    
    it('should disable input', async () => {
        const { dom } = await setupPickerInputForTest({
            value: undefined,
            selectionMode: 'single',
            isDisabled: true,
        });
        
        expect(dom.input?.hasAttribute('disabled')).toBeTruthy();
        expect(dom.input?.getAttribute('aria-disabled')?.trim()).toEqual('true');

        fireEvent.click(dom.input as Element);
        expect(screen.queryByRole('dialog')).toBeNull();
    });

    it('should make an input readonly', async () => {
        const { dom } = await setupPickerInputForTest({
            value: undefined,
            selectionMode: 'single',
            isReadonly: true,
        });
        
        expect(dom.input?.hasAttribute('readonly')).toBeTruthy();
        expect(dom.input?.getAttribute('aria-readonly')?.trim()).toEqual('true');

        fireEvent.click(dom.input as Element);
        expect(screen.queryByRole('dialog')).toBeNull();
    });
    
    it('[selectionMode single] should render names of items by getName', async () => {
        const { mocks, dom } = await setupPickerInputForTest({
            value: 3,
            selectionMode: 'single',
            getName: ({ name }) => name,
        });

        expect(dom.input?.getAttribute('placeholder')?.trim()).toBeUndefined();
        
        await delayAct(100);
       
        expect(dom.input?.getAttribute('placeholder')?.trim()).toEqual(languageLevels[1].name);
        
        fireEvent.click(dom.input as Element);
        expect(screen.getByRole('dialog')).toBeInTheDocument();
        const optionC2 = await screen.findByText('Proficiency');
        fireEvent.click(optionC2);
        expect(mocks.onValueChange).toHaveBeenLastCalledWith(12);
    });
    
    it('[selectionMode multi] should render names of items by getName', async () => {
        const { dom } = await setupPickerInputForTest({
            value: [3, 4],
            selectionMode: 'multi',
            getName: ({ name }) => name,
        });

        const selectedItemsNames1 = screen.queryAllByTestId(/uui-PickerToggler-item/)
            .map((button) => button.textContent?.trim());
        
        expect(selectedItemsNames1).toEqual(['', '']);

        expect(dom.input?.getAttribute('placeholder')?.trim()).toEqual('Please select');
   
        await delayAct(100);

        const selectedItemsNames2 = screen.queryAllByTestId(/uui-PickerToggler-item/)
            .map((button) => button.textContent?.trim());
        
        expect(selectedItemsNames2).toEqual(['Elementary+', 'Pre-Intermediate']);
    });

    it('[selectionMode single] should render entity name in placeholder', async () => {
        const { dom } = await setupPickerInputForTest({
            value: undefined,
            selectionMode: 'single',
            entityName: 'Language Level',
        });
        
        expect(dom.input?.getAttribute('placeholder')?.trim()).toEqual('Please select Language Level');
    });
    
    it('[selectionMode single] should ignore plural entity name in placeholder', async () => {
        const { dom } = await setupPickerInputForTest({
            value: undefined,
            selectionMode: 'single',
            entityName: 'Language Level',
            entityPluralName: 'Multiple Language Levels',
        });
        
        expect(dom.input?.getAttribute('placeholder')?.trim()).toEqual('Please select Language Level');
    });

    it('[selectionMode multi] should render entity name with \'s\' in placeholder', async () => {
        const { dom } = await setupPickerInputForTest({
            value: undefined,
            selectionMode: 'multi',
            entityName: 'Language Level',
        });
        
        expect(dom.input?.getAttribute('placeholder')?.trim()).toEqual('Please select Language Levels');
    });

    it('[selectionMode multi] should render plural entity name in placeholder', async () => {
        const { dom } = await setupPickerInputForTest({
            value: undefined,
            selectionMode: 'multi',
            entityName: 'Language Level',
            entityPluralName: 'Multiple Language Levels',
        });
        
        expect(dom.input?.getAttribute('placeholder')?.trim()).toEqual('Please select Multiple Language Levels');
    });

    it.each<[CascadeSelection]>(
        [[false], [true], ['implicit'], ['explicit']],
    )
    ('[selectionMode single] should pick single element with cascadeSelection = %s', async (cascadeSelection) => {
        const { mocks, dom } = await setupPickerInputForTest({
            value: undefined,
            getName: ({ name }) => name,
            selectionMode: 'single',
            cascadeSelection,
            dataSource: mockTreeLikeDataSourceAsync,
        });

        fireEvent.click(dom.input as HTMLElement);
        await delayAct(100);

        const [, second] = within(screen.getByRole('dialog')).queryAllByTestId(/uui-PickerInput-item/);
        fireEvent.click(second);

        expect(mocks.onValueChange).toHaveBeenLastCalledWith(2);
        expect(dom.input?.getAttribute('placeholder')?.trim()).toEqual('Parent 2');
    });
    
    it('[selectionMode multi] should pick single element with cascadeSelection = false', async () => {
        const { mocks, dom } = await setupPickerInputForTest({
            value: undefined,
            getName: ({ name }) => name,
            selectionMode: 'multi',
            cascadeSelection: false,
            dataSource: mockTreeLikeDataSourceAsync,
        });

        fireEvent.click(dom.input as HTMLElement);
        await delayAct(100);

        const [, second] = within(screen.getByRole('dialog')).queryAllByTestId(/uui-PickerInput-item/);
        const checkbox = await within(second).findByRole('checkbox');
        fireEvent.click(checkbox);

        expect(mocks.onValueChange).toHaveBeenLastCalledWith([2]);
        const selectedItemsNames1 = screen.queryAllByTestId(/uui-PickerToggler-item/)
            .map((button) => button.textContent?.trim());
        
        expect(checkbox).toBeChecked();
        expect(selectedItemsNames1).toEqual(['Parent 2']);
    });
});
