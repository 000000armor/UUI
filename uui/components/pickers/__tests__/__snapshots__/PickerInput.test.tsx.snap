// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PickerInput should render icon at specific position 1`] = `"<div class=\\"container uui-input-box -clickable root size-36 mode-form\\" tabindex=\\"0\\"><div class=\\"body multiline\\"><div class=\\"container uui-icon uui-enabled\\"><div data-testid=\\"test-icon\\"></div></div><input type=\\"text\\" tabindex=\\"-1\\" aria-haspopup=\\"true\\" autocomplete=\\"no\\" class=\\"uui-input cursorText\\" placeholder=\\"Please select \\" readonly=\\"\\" value=\\"\\"></div><div class=\\"actions\\"><div class=\\"container uui-icon uui-enabled -clickable uui-icon-dropdown\\"><svg class=\\"\\"></svg></div></div></div>"`;

exports[`PickerInput should render icon at specific position 2`] = `"<div class=\\"container uui-input-box -clickable root size-36 mode-form\\" tabindex=\\"0\\"><div class=\\"body multiline\\"><div class=\\"container uui-icon uui-enabled\\"><div data-testid=\\"test-icon\\"></div></div><input type=\\"text\\" tabindex=\\"-1\\" aria-haspopup=\\"true\\" autocomplete=\\"no\\" class=\\"uui-input cursorText\\" placeholder=\\"Please select \\" readonly=\\"\\" value=\\"\\"></div><div class=\\"actions\\"><div class=\\"container uui-icon uui-enabled -clickable uui-icon-dropdown\\"><svg class=\\"\\"></svg></div></div></div>"`;

exports[`PickerInput should render icon at specific position 3`] = `"<div class=\\"container uui-input-box -clickable root size-36 mode-form\\" tabindex=\\"0\\"><div class=\\"body multiline\\"><input type=\\"text\\" tabindex=\\"-1\\" aria-haspopup=\\"true\\" autocomplete=\\"no\\" class=\\"uui-input cursorText\\" placeholder=\\"Please select \\" readonly=\\"\\" value=\\"\\"><div class=\\"container uui-icon uui-enabled\\"><div data-testid=\\"test-icon\\"></div></div></div><div class=\\"actions\\"><div class=\\"container uui-icon uui-enabled -clickable uui-icon-dropdown\\"><svg class=\\"\\"></svg></div></div></div>"`;

exports[`PickerInput should render input as invalid 1`] = `"<div class=\\"container uui-input-box -clickable root size-36 mode-form\\" tabindex=\\"0\\"><div class=\\"body\\"><input type=\\"text\\" tabindex=\\"-1\\" aria-haspopup=\\"true\\" autocomplete=\\"no\\" class=\\"uui-input singleInput cursorText\\" placeholder=\\"Please select \\" readonly=\\"\\" value=\\"\\"></div><div class=\\"actions\\"><div class=\\"container uui-icon uui-enabled -clickable uui-icon-dropdown\\"><svg class=\\"\\"></svg></div></div></div>"`;

exports[`PickerInput should render input as invalid 2`] = `"<div class=\\"container uui-input-box uui-invalid -clickable root size-36 mode-form\\" tabindex=\\"0\\"><div class=\\"body\\"><input type=\\"text\\" tabindex=\\"-1\\" aria-haspopup=\\"true\\" autocomplete=\\"no\\" class=\\"uui-input singleInput cursorText\\" placeholder=\\"Please select \\" readonly=\\"\\" value=\\"\\"></div><div class=\\"actions\\"><div class=\\"container uui-icon uui-enabled -clickable uui-icon-dropdown\\"><svg class=\\"\\"></svg></div></div></div>"`;

exports[`PickerInput should render with maximum props 1`] = `
<div
  className="container uui-input-box -clickable root size-48 mode-form"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  tabIndex={0}
>
  <div
    className="body multiline"
  >
    <button
      className="container uui-button-box uui-enabled size-42 root"
      onClick={[Function]}
      tabIndex={-1}
      type="button"
    >
      <div
        className="uui-caption"
      >
        A1
      </div>
      <div
        className="container uui-icon uui-enabled -clickable -clickable"
        onClick={[Function]}
        style={Object {}}
      >
        <svg
          className=""
        />
      </div>
    </button>
    <button
      className="container uui-button-box uui-enabled size-42 root"
      onClick={[Function]}
      tabIndex={-1}
      type="button"
    >
      <div
        className="uui-caption"
      >
        A1+
      </div>
      <div
        className="container uui-icon uui-enabled -clickable -clickable"
        onClick={[Function]}
        style={Object {}}
      >
        <svg
          className=""
        />
      </div>
    </button>
  </div>
  <div
    className="actions"
  >
    <div
      aria-label="Clear"
      className="container uui-icon uui-enabled -clickable uui-icon-cancel -clickable"
      onClick={[Function]}
      role="button"
      style={Object {}}
      tabIndex={-1}
    >
      <svg
        className=""
      />
    </div>
    <div
      className="container uui-icon uui-enabled -clickable uui-icon-dropdown"
      onClick={[Function]}
      style={Object {}}
    >
      <svg
        className=""
      />
    </div>
  </div>
</div>
`;

exports[`PickerInput should render with minimum props 1`] = `
<div
  className="container uui-input-box -clickable root size-36 mode-form"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  tabIndex={0}
>
  <div
    className="body"
  >
    <input
      aria-haspopup={true}
      autoComplete="no"
      className="uui-input singleInput cursorText"
      onChange={[Function]}
      placeholder="Please select "
      readOnly={true}
      tabIndex={-1}
      type="text"
      value=""
    />
  </div>
  <div
    className="actions"
  >
    <div
      className="container uui-icon uui-enabled -clickable uui-icon-dropdown"
      onClick={[Function]}
      style={Object {}}
    >
      <svg
        className=""
      />
    </div>
  </div>
</div>
`;

exports[`PickerInput should render with mode = cell 1`] = `"<div class=\\"container uui-input-box -clickable root size-36 mode-cell\\" tabindex=\\"0\\"><div class=\\"body multiline\\"><input type=\\"text\\" tabindex=\\"-1\\" aria-haspopup=\\"true\\" autocomplete=\\"no\\" class=\\"uui-input cursorText\\" placeholder=\\"Please select \\" readonly=\\"\\" value=\\"\\"></div><div class=\\"actions\\"><div class=\\"container uui-icon uui-enabled -clickable uui-icon-dropdown\\"><svg class=\\"\\"></svg></div></div></div>"`;

exports[`PickerInput should render with mode = form 1`] = `"<div class=\\"container uui-input-box -clickable root size-36 mode-form\\" tabindex=\\"0\\"><div class=\\"body multiline\\"><input type=\\"text\\" tabindex=\\"-1\\" aria-haspopup=\\"true\\" autocomplete=\\"no\\" class=\\"uui-input cursorText\\" placeholder=\\"Please select \\" readonly=\\"\\" value=\\"\\"></div><div class=\\"actions\\"><div class=\\"container uui-icon uui-enabled -clickable uui-icon-dropdown\\"><svg class=\\"\\"></svg></div></div></div>"`;

exports[`PickerInput should render with mode = inline 1`] = `"<div class=\\"container uui-input-box -clickable root size-36 mode-inline\\" tabindex=\\"0\\"><div class=\\"body multiline\\"><input type=\\"text\\" tabindex=\\"-1\\" aria-haspopup=\\"true\\" autocomplete=\\"no\\" class=\\"uui-input cursorText\\" placeholder=\\"Please select \\" readonly=\\"\\" value=\\"\\"></div><div class=\\"actions\\"><div class=\\"container uui-icon uui-enabled -clickable uui-icon-dropdown\\"><svg class=\\"\\"></svg></div></div></div>"`;

exports[`PickerInput should render with mode = undefined 1`] = `"<div class=\\"container uui-input-box -clickable root size-36 mode-form\\" tabindex=\\"0\\"><div class=\\"body multiline\\"><input type=\\"text\\" tabindex=\\"-1\\" aria-haspopup=\\"true\\" autocomplete=\\"no\\" class=\\"uui-input cursorText\\" placeholder=\\"Please select \\" readonly=\\"\\" value=\\"\\"></div><div class=\\"actions\\"><div class=\\"container uui-icon uui-enabled -clickable uui-icon-dropdown\\"><svg class=\\"\\"></svg></div></div></div>"`;

exports[`PickerInput should support single line 1`] = `"<div class=\\"container uui-input-box -clickable root size-36 mode-form\\" tabindex=\\"0\\"><div class=\\"body multiline\\"><input type=\\"text\\" tabindex=\\"-1\\" aria-haspopup=\\"true\\" autocomplete=\\"no\\" class=\\"uui-input cursorText\\" placeholder=\\"Please select \\" readonly=\\"\\" value=\\"\\"></div><div class=\\"actions\\"><div class=\\"container uui-icon uui-enabled -clickable uui-icon-dropdown\\"><svg class=\\"\\"></svg></div></div></div>"`;

exports[`PickerInput should support single line 2`] = `"<div class=\\"container uui-input-box -clickable root size-36 mode-form\\" tabindex=\\"0\\"><div class=\\"body\\"><input type=\\"text\\" tabindex=\\"-1\\" aria-haspopup=\\"true\\" autocomplete=\\"no\\" class=\\"uui-input cursorText\\" placeholder=\\"Please select \\" readonly=\\"\\" value=\\"\\"></div><div class=\\"actions\\"><div class=\\"container uui-icon uui-enabled -clickable uui-icon-dropdown\\"><svg class=\\"\\"></svg></div></div></div>"`;
